internal/cmd/changelog/cmd/generate.go

```diff
@@ -147,34 +147,118 @@ func findPreviousRelease(r *git.Repository, v *semver.Version) (*semver.Version,
 
 // findNewCommits queries the git repository for new commits since the last tagged release.
 func findNewCommits(r *git.Repository, prevHash, currHash plumbing.Hash) ([]Commit, error) {
-	logs, err := r.Log(&git.LogOptions{
-		From: currHash,
-	})
+	mergeBase, err := findMergeBase(r, prevHash, currHash)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to load logs from %v", currHash)
+		return nil, err
 	}
-	defer logs.Close()
+
+	queue := []plumbing.Hash{currHash}
+	seen := make(map[plumbing.Hash]bool)
 
 	var commits []Commit
-	for {
-		commit, err := logs.Next()
+	for len(queue) > 0 {
+		h := queue[0]
+		queue = queue[1:]
+
+		if seen[h] {
+			continue
+		}
+		seen[h] = true
+
+		if h == prevHash || h == mergeBase {
+			// Do not process this commit or its
+			// parents. We are done processing commits on this
+			// branch.
+			continue
+		}
+
+		commit, err := r.CommitObject(h)
 		if err != nil {
-			if err == io.EOF {
-				break
-			}
 			return nil, err
 		}
-		if commit.Hash == prevHash {
-			break
+
+		// Ignore merge commits when parsing commit messages.
+		if len(commit.ParentHashes) <= 1 {
+			c, err := parseCommit(commit)
+			if err == nil {
+				commits = append(commits, c)
+			} else {
+				log.Printf("invalid commit message %v: %v", commit.Hash, err)
+			}
 		}
-		c, err := parseCommit(commit)
-		if err == nil {
-			commits = append(commits, c)
+
+		// Iterate over the parents and add them to the queue.
+		queue = append(queue, commit.ParentHashes...)
+	}
+	return commits, nil
+}
+
+// findMergeBase finds the common parent commit the two commits have.
+func findMergeBase(r *git.Repository, prevHash, currHash plumbing.Hash) (plumbing.Hash, error) {
+	prevLog, err := r.Log(&git.LogOptions{
+		From: prevHash,
+	})
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+	defer prevLog.Close()
+
+	prevCommit, err := prevLog.Next()
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+
+	currLog, err := r.Log(&git.LogOptions{
+		From: currHash,
+	})
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+	defer currLog.Close()
+
+	currCommit, err := currLog.Next()
+	if err != nil {
+		return plumbing.Hash{}, err
+	}
+
+	seen := map[plumbing.Hash]bool{
+		prevCommit.Hash: true,
+		currCommit.Hash: true,
+	}
+	for {
+		var (
+			commit **object.Commit
+			log    object.CommitIter
+		)
+
+		if prevCommit == nil {
+			commit, log = &currCommit, currLog
+		} else if currCommit == nil {
+			commit, log = &prevCommit, prevLog
 		} else {
-			log.Printf("invalid commit message %v: %v", commit.Hash, err)
+			if currCommit.Committer.When.Before(prevCommit.Committer.When) {
```

Comment:
Is this `When` field that actual time that the commit was added relative to other commits?

I have seen dates in the git log that are definitely from way before the commit was added, which must be due to `git commit --amend` or force push.

Reply:
It is. Git is... not very smart about this. It's what the actual `merge-base` command does. Here's the code that sets the comparison:

https://github.com/git/git/blob/master/commit-reach.c#L38

The generation that's referenced comes from a commit graph that isn't written to disk unless you explicitly write it to disk so it's almost always empty and isn't exposed by the go-git library.

The generation number would be much better and exactly what would be needed, but it doesn't exist for practical purposes. Git code just seems to use commit dates as a way of trying to reduce the amount of time for practical purposes, but it's obviously not very reliable.